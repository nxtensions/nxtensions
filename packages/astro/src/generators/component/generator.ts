import { formatFiles, Tree } from '@nx/devkit';
import type { GeneratorOptions } from './schema';
import {
  addComponentFile,
  addStyleDependencies,
  normalizeOptions,
} from './utilities';

export async function componentGenerator(
  tree: Tree,
  rawOptions: GeneratorOptions
) {
  const options = normalizeOptions(tree, rawOptions);

  addComponentFile(tree, options);
  const styleTask = addStyleDependencies(tree, options);
  await formatFiles(tree);

  return styleTask;
}

export default componentGenerator;
