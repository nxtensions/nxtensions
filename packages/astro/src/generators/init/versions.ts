export const astroVersion = '~1.0.0-beta.56';
