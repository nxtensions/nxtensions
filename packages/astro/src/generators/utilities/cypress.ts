export async function importNrwlCypress(): Promise<
  typeof import('@nx/cypress')
> {
  try {
    return await import('@nx/cypress');
  } catch {
    throw new Error(
      `The "@nx/cypress" package is missing and it's required to generate an E2E project. Please install it and try again.`
    );
  }
}
