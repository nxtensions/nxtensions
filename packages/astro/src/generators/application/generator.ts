import {
  ensurePackage,
  formatFiles,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';
import { addDependenciesToPackageJson } from '../../utilities/package-json';
import { initGenerator } from '../init/generator';
import {
  astroCheckVersion,
  astroVersion,
  getInstalledNxVersion,
  typescriptVersion,
} from '../utilities/versions';
import type { GeneratorOptions } from './schema';
import {
  addFiles,
  addIntegrationsPackages,
  addProject,
  configureTailwindIntegration,
  normalizeOptions,
  setDefaultProject,
  setupE2ETests,
} from './utilities';

export async function applicationGenerator(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback | void> {
  return await applicationGeneratorInternal(tree, {
    projectNameAndRootFormat: 'derived',
    ...rawOptions,
  });
}

export async function applicationGeneratorInternal(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback | void> {
  const options = await normalizeOptions(tree, rawOptions);

  const tasks: GeneratorCallback[] = [];
  const { initGenerator: jsInitGenerator } = ensurePackage<
    typeof import('@nx/js')
  >('@nx/js', getInstalledNxVersion(tree));
  const jsTask = await jsInitGenerator(tree, { skipFormat: true });
  tasks.push(jsTask);

  await initGenerator(tree, {});

  const depsTask = addDependenciesToPackageJson(
    tree,
    {},
    {
      astro: astroVersion,
      '@astrojs/check': astroCheckVersion,
      typescript: typescriptVersion,
    }
  );
  tasks.push(depsTask);

  addProject(tree, options);
  setDefaultProject(tree, options.projectName);
  addFiles(tree, options);
  await configureTailwindIntegration(tree, options);

  const e2eTask = await setupE2ETests(tree, options);
  if (e2eTask) {
    tasks.push(e2eTask);
  }

  const integrationsTask = addIntegrationsPackages(tree, options.integrations);
  if (integrationsTask) {
    tasks.push(integrationsTask);
  }

  await formatFiles(tree);

  return () => tasks.forEach((task) => task());
}

export default applicationGenerator;
