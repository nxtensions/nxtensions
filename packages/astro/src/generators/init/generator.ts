import { cypressInitGenerator } from '@nrwl/cypress';
import {
  addDependenciesToPackageJson,
  GeneratorCallback,
  Tree,
} from '@nrwl/devkit';
import { GeneratorOptions } from './schema';
import {
  addProjectGraphPlugin,
  updateGitignore,
  addVSCodeRecommendedExtensions,
} from './utilities';
import { astroVersion } from './versions';

export function initGenerator(
  tree: Tree,
  options: GeneratorOptions
): GeneratorCallback {
  addProjectGraphPlugin(tree);
  updateGitignore(tree);
  addVSCodeRecommendedExtensions(tree);

  const tasks: GeneratorCallback[] = [];
  if (options.addCypressTests !== false) {
    const cypressTask = cypressInitGenerator(tree);
    tasks.push(cypressTask);
  }

  const depsTask = addDependenciesToPackageJson(
    tree,
    {},
    { astro: astroVersion }
  );
  tasks.push(depsTask);

  return () => tasks.forEach((task) => task());
}

export default initGenerator;
