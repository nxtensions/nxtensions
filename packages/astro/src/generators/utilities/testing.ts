import { Tree, readJsonFile, updateJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace as _createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { PackageJson } from '../../utilities/package-json';

export function createTreeWithEmptyWorkspace(nxVersion?: string): Tree {
  const tree = _createTreeWithEmptyWorkspace();
  if (!nxVersion) {
    const { devDependencies } = readJsonFile<PackageJson>('package.json');
    nxVersion = devDependencies['nx'];
  }
  updateJson(tree, 'package.json', (json) => {
    json.devDependencies['nx'] = nxVersion;
    return json;
  });

  return tree;
}
