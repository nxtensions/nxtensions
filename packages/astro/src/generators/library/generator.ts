import { formatFiles, GeneratorCallback, Tree } from '@nrwl/devkit';
import { initGenerator } from '../init/generator';
import { GeneratorOptions } from './schema';
import {
  addFiles,
  addPathMapping,
  addProject,
  normalizeOptions,
} from './utilities';

export async function libraryGenerator(
  tree: Tree,
  rawOptions: GeneratorOptions
): Promise<GeneratorCallback> {
  if (rawOptions.publishable && !rawOptions.importPath) {
    throw new Error(
      'For publishable libraries the "--importPath" must be provided. Please note it needs to be a valid npm package name (e.g. my-lib or @my-org/my-lib).'
    );
  }

  const options = normalizeOptions(tree, rawOptions);

  const initTask = initGenerator(tree, { addCypressTests: false });

  addProject(tree, options);
  addFiles(tree, options);
  addPathMapping(tree, options);

  await formatFiles(tree);

  return initTask;
}

export default libraryGenerator;
