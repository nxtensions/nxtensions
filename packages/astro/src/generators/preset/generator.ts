import {
  formatFiles,
  joinPathFragments,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';
import { applicationGenerator } from '../application/generator';
import type { GeneratorOptions } from './schema';

export default async function (
  tree: Tree,
  options: GeneratorOptions
): Promise<GeneratorCallback | void> {
  const applicationTask = await applicationGenerator(tree, {
    name: options.astroAppName,
    tags: options.tags,
    directory: joinPathFragments('apps', options.astroAppName),
    projectNameAndRootFormat: 'as-provided',
  });

  await formatFiles(tree);

  return applicationTask;
}
