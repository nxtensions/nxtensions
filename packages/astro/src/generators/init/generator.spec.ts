jest.mock('@nrwl/cypress');

import { cypressInitGenerator } from '@nrwl/cypress';
import {
  readJson,
  readWorkspaceConfiguration,
  Tree,
  updateJson,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { initGenerator } from './generator';
import { astroVersion } from './versions';

describe('init generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace(2);
    jest.clearAllMocks();
  });

  test('should add project graph plugin', () => {
    initGenerator(tree, {});

    const { plugins } = readWorkspaceConfiguration(tree);
    expect(plugins).toContain('@nxtensions/astro');
  });

  test('should add the check target to the cacheable operations', async () => {
    initGenerator(tree, {});

    const workspace = readWorkspaceConfiguration(tree);
    expect(
      workspace.tasksRunnerOptions.default.options.cacheableOperations
    ).toContain('check');
  });

  test('should correct node_modules entry in .gitignore file when only targeting root', () => {
    tree.write(
      '.gitignore',
      `foo
bar

/node_modules
`
    );

    initGenerator(tree, {});

    const gitignore = tree.read('.gitignore', 'utf-8');
    expect(gitignore).toBe(`foo
bar

node_modules
`);
  });

  test('should add node_modules entry to .gitignore', () => {
    tree.write(
      '.gitignore',
      `foo
bar
`
    );

    initGenerator(tree, {});

    const gitignore = tree.read('.gitignore', 'utf-8');
    expect(gitignore).toBe(`foo
bar

node_modules
`);
  });

  test('should not add VSCode extensions file when it does not exist', async () => {
    initGenerator(tree, {});

    expect(tree.exists('.vscode/extensions.json')).toBeFalsy();
  });

  test('should add the Astro extension to the VSCode recommended extensions', async () => {
    tree.write(
      '.vscode/extensions.json',
      JSON.stringify({
        recommendations: ['nrwl.angular-console'],
      })
    );

    initGenerator(tree, {});

    const { recommendations } = readJson(tree, '.vscode/extensions.json');
    expect(recommendations).toContain('astro-build.astro-vscode');
  });

  test('should not duplicate the Astro extension when it is already present', async () => {
    tree.write(
      '.vscode/extensions.json',
      JSON.stringify({
        recommendations: ['nrwl.angular-console', 'astro-build.astro-vscode'],
      })
    );

    initGenerator(tree, {});

    const { recommendations } = readJson(tree, '.vscode/extensions.json');
    expect(recommendations).toStrictEqual([
      'nrwl.angular-console',
      'astro-build.astro-vscode',
    ]);
  });

  test('should add astro as a devDependency', () => {
    initGenerator(tree, {});

    const { devDependencies } = readJson(tree, 'package.json');
    expect(devDependencies.astro).toBe(astroVersion);
  });

  test('should generate script to patch the nx cli', () => {
    initGenerator(tree, {});

    expect(tree.exists('tools/scripts/patch-nx-cli.js')).toBe(true);
  });

  test('should add a postinstall script to patch the nx cli', () => {
    initGenerator(tree, {});

    const { scripts } = readJson(tree, 'package.json');
    expect(scripts.postinstall).toBe('node ./tools/scripts/patch-nx-cli.js');
  });

  test('should update the postinstall script to patch the nx cli', () => {
    updateJson(tree, 'package.json', (json) => ({
      ...json,
      scripts: { postinstall: 'node ./some-important-script.js' },
    }));

    initGenerator(tree, {});

    const { scripts } = readJson(tree, 'package.json');
    expect(scripts.postinstall).toBe(
      'node ./some-important-script.js && node ./tools/scripts/patch-nx-cli.js'
    );
  });

  describe('--addCypressTests', () => {
    test('should not add cypress when --addCypressTests=false', () => {
      initGenerator(tree, { addCypressTests: false });

      expect(cypressInitGenerator).not.toHaveBeenCalled();
    });

    test('should add cypress by default', () => {
      initGenerator(tree, {});

      expect(cypressInitGenerator).toHaveBeenCalled();
    });

    test('should add cypress when --addCypressTests=true', () => {
      initGenerator(tree, {});

      expect(cypressInitGenerator).toHaveBeenCalled();
    });
  });
});
