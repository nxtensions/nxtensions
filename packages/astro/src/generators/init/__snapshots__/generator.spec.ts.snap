// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init generator should add .npmrc when it does not exist 1`] = `
"# Expose Astro dependencies for \\\`pnpm\\\` users
shamefully-hoist=true"
`;

exports[`init generator should update .npmrc when it exist 1`] = `
"# Some initial content
# Expose Astro dependencies for \\\`pnpm\\\` users
shamefully-hoist=true"
`;
